programs/ido-program/programs/ido-program/src/lib.rs

```diff
@@ -0,0 +1,438 @@
+        let cpi_ctx = CpiContext::new_with_signer(cpi_program, cpi_accounts, signer);
+
+        token::transfer(cpi_ctx, native_amount as u64)?;
+
+        Ok(())
+    }
+
+    #[access_control(can_withdraw_fiat(&ctx.accounts.pool, &ctx.accounts.clock))]
+        pub fn withdraw_pool_fiat(ctx: Context<WithdrawPoolFiat>) -> Result<()> {            
+            let seeds = &[
+                ctx.accounts.pool.native_mint.as_ref(),
+                &[ctx.accounts.pool.bump],
+            ];
+            let signer = &[&seeds[..]];
+            let cpi_accounts = Transfer {
+                from: ctx.accounts.pool_fiat.to_account_info(),
+                to: ctx.accounts.creator_fiat.to_account_info(),
+                authority: ctx.accounts.pool_signer.to_account_info(),
+            };
+            let cpi_program = ctx.accounts.token_program.to_account_info().clone();
+            let cpi_ctx = CpiContext::new_with_signer(cpi_program, cpi_accounts, signer);
+            token::transfer(cpi_ctx, ctx.accounts.pool_fiat.amount)?;
+
+            Ok(())
+        }
+}
+
+
+
+#[derive(Accounts)]
+pub struct InitializePool<'info> {
+    #[account(init, payer = authority, space = PoolAccount::LEN)]
+    pub pool: Box<Account<'info, PoolAccount>>,
+
+    /// CHECK: This is not dangerous
+    #[account(mut)]
+    pub pool_signer: AccountInfo<'info>,
+
+    #[account(
+        constraint = redeemable_mint.mint_authority == COption::Some(*pool_signer.key), 
+        constraint = redeemable_mint.supply == 0
+    )]
+    pub redeemable_mint: Box<Account<'info, Mint>>,
+
+    #[account(constraint = fiat_mint.decimals == redeemable_mint.decimals)]
+    pub fiat_mint: Box<Account<'info, Mint>>,
+
+    #[account(constraint = pool_native.mint == *native_mint.to_account_info().key)]
+    pub native_mint: Box<Account<'info, Mint>>,
+
+    #[account(mut, constraint = pool_native.owner == *pool_signer.key)]
+    pub pool_native: Box<Account<'info, TokenAccount>>,
+
+    #[account(constraint = pool_fiat.owner == *pool_signer.key)]
+    pub pool_fiat: Box<Account<'info, TokenAccount>>,
+
+    #[account(mut)]
+    pub authority: Signer<'info>,
+
+    #[account(mut)]
+    pub creator_native: Box<Account<'info, TokenAccount>>,
+
+    #[account(constraint = token_program.key == &token::ID)]
+    pub token_program: Program<'info, Token>,
+
+    pub rent: Sysvar<'info, Rent>,
+
+    pub clock: Sysvar<'info, Clock>,
+
+    pub system_program: Program<'info, System>,
+}
+
+impl<'info> InitializePool<'info> {
+    fn accounts(ctx: &Context<InitializePool<'info>>, bump: u8) -> Result<()> {
+        let expected_signer = Pubkey::create_program_address(
+            &[ctx.accounts.pool_native.mint.as_ref(), &[bump]],
+            ctx.program_id,
+        )
+        .map_err(|_| ErrorCode::InvalidBump)?;
+        if ctx.accounts.pool_signer.key != &expected_signer {
+            return Err(ErrorCode::InvalidBump.into());
+        }
+        Ok(())
+    }
+}
```

Comment:
What is the purpose of this when you could use `#[account(seeds = ...)`?

Reply:
The bump is coming as the instruction's function parameter and we are using `Pubkey::create_program_address()` to check that the `pool_signer` sent in the accounts struct is correct or not.

Reply:
Though both are valid and you are free to use Native validations, it's not "Anchor way" of doing things.

There is `#[instruction(...)]` macro if you want to get instruction arguments and validating seeds/bump should be done with `#[account(...)]` macro, especially in the Anchor Book.